rpm/0070-sailfishos-webrtc-Implement-video-capture-module.-JB.patch

```diff
@@ -0,0 +1,2255 @@
++        const char* /*deviceUniqueIdUTF8*/,
++        const char* /*dialogTitleUTF8*/,
++        void* /*parentWindow*/,
++        uint32_t /*positionX*/,
++        uint32_t /*positionY*/) { return -1; }
++    void FillCapabilities(const char* devName);
++    int32_t Init();
++private:
++    void FindCameras();
++    std::vector<CameraInfo> _deviceList;
++    std::map<std::string, camera_type_t> _deviceMap;
++    static std::vector<camera_type_t> supportedCameraTypes();
++};
++}  // namespace videocapturemodule
++}  // namespace webrtc
++#endif // WEBRTC_MODULES_VIDEO_CAPTURE_MAIN_SOURCE_SFOS_DEVICE_INFO_SFOS_H_
+diff --git a/media/webrtc/trunk/webrtc/modules/video_capture/sfos/gst_droid_camera.cc b/media/webrtc/trunk/webrtc/modules/video_capture/sfos/gst_droid_camera.cc
+new file mode 100644
+index 000000000000..1d6c4f012005
+--- /dev/null
++++ b/media/webrtc/trunk/webrtc/modules/video_capture/sfos/gst_droid_camera.cc
+@@ -0,0 +1,496 @@
++/*
++ *  Copyright (c) 2021 Open Mobile Platform LLC. All Rights Reserved.
++ *
++ *  Use of this source code is governed by a BSD-style license
++ *  that can be found in the LICENSE file in the root of the source
++ *  tree. An additional intellectual property rights grant can be found
++ *  in the file PATENTS.  All contributing project authors may
++ *  be found in the AUTHORS file in the root of the source tree.
++ */
++
++#include <errno.h>
++#include <fcntl.h>
++#include <stdio.h>
++#include <string.h>
++#include <sys/ioctl.h>
++#include <sys/mman.h>
++#include <sys/stat.h>
++#include <unistd.h>
++
++#include <new>
++
++#include "webrtc/base/refcount.h"
++#include "webrtc/base/scoped_ref_ptr.h"
++#include "webrtc/system_wrappers/include/trace.h"
++
++#include "webrtc/modules/video_capture/sfos/gst_droid_camera.h"
++
++namespace webrtc {
++namespace videocapturemodule {
++
++#define EXPECTED_CAPTURE_DELAY_PREF "media.getusermedia.camera.expected_capture_delay"
++
++#define GST_DROID_CAMERA_PRIMARY    "primary"
++#define GST_DROID_CAMERA_SECONDARY  "secondary"
++#define GST_DROID_CAMERA_VENDOR     "gst-droid"
++
++GstDroidCamera::GstDroidCamera()
++{
++    LOG_F(LS_VERBOSE);
++}
++
++int32_t GstDroidCamera::Init(const char* deviceUniqueIdUTF8)
++{
++    GstElement *sink;
++    int32_t ret = -1;
++    int cameraNr;
++
++    LOG_F(LS_INFO) << deviceUniqueIdUTF8;
++
++    if (!gst_is_initialized ()) {
++        gst_init(NULL, NULL);
++    }
++
++    _captureContext = g_main_context_new();
++    if (!_captureContext) {
++        return -1;
++    }
++
++    g_main_context_push_thread_default (_captureContext);
++
++    /* TODO: Make the pipeline configurable */
++    _pipeline = gst_parse_launch (
++        "droidcamsrc mode=2 exposure-mode=auto iso_speed=0 name=src"
++        " src.vfsrc"
++        " ! capsfilter name=camera_capsfilter"
++        " ! videorate drop-only=true"
++        " ! capsfilter name=rate_capsfilter"
++        " ! queue ! appsink name=videosource", NULL );
```

Comment:
I realise build repo dependencies make this impossible at the moment, but if it could be resolved so that droidmedia is available to use directly would that be an option?

It's not necessarily going to be any easier to use, but it may open up some option to supply this API's buffer type to the camera API and eliminate a copy.

Reply:
The gecko has some support for native buffers, so (after a massive patching and rewriting) it can pass native buffers directly to the encoder. It's my goal for the future. I tried to create a capture module based on plain droidmedia: although it calls nearly the same droidmedia functions as its gst version, I've been unable to get it to produce buffers. Perhaps I'm missing something.

Reply:
What do you mean by native buffers here? Because the android camera has two optimized buffer types it outputs. There are the preview stream graphics memory buffers (gralloc) which can be consumed by the GPU directly without any memory  copies. And there are the recording stream metadata headers which can be consumed by the hardware accelerated video encoders. And just to complicate things further there is a way to connect the recording stream to the video encoder in the android media server so the application doesn't see the video data until after it is encoded, this is what gst-droid utilizes currently and although there is an option to do the data passing in the application my understanding is that it doesn't currently and perhaps never worked.

I don't think there's any trick to using droidmedia directly at least as far as getting preview frames is concerned. It should be a matter of creating the camera object, locking it, setting up callbacks on both the camera and its buffer queue, then starting the preview stream. Then graphics buffers should start arriving in the buffer queue callbacks. There are also some flags you can set to get raw data in the preview_frame_cb callback, although that's only valid until the callback returns so there's a necessary copy to work with that most likely.

This is what I've tried as a standalone thing https://gist.github.com/adenexter/89375f20793fcee698d623597b84ff0e. It's not displaying anything but you can at least see it's querying the cameras and starts one and is continually receiving frames. 

On a Jolla C, and a T8 it's producing both preview and recording frames, but on an XA2 the recording stream doesn't work and as seen in David's screenshot the color format of the unlocked graphics buffers is not YV12 as assumed but is instead HAL_PIXEL_FORMAT_IMPLEMENTATION_DEFINED. In the parameters the preview-format is listed as yuv420sp as it is on the other devices but at I guess only the data in the preview_frame_cb callback is in that format. It can probably be fixed in the gstreamer implementation by dropping the memory:DroidMediaQueueBuffer feature from the caps negotiation, probably at the cost of another copy.

Reply:
By native handle I mean graphics memory buffers. I was thinking about passing a handle to such buffer to the GMP encoder plugin instead of managing shared memory and copying data back and fourth.

Thank you for giving the example of using droidmedia. Your code works, but I still can't see the difference with mine. Interesting, will look closer at this.

I was able to repeat David's result at my Sony 10. I'm using viewfinder frames in my capture code. Can't prove it but I think the data is in RGB format and perhaps that's due to the buffer's usage [here](https://github.com/sailfishos/droidmedia/blob/766eb45a873dee3941eb4d455a92544eddf78c4f/private.cpp#L86): 

m_queue->setConsumerUsageBits(android::GraphicBuffer::USAGE_HW_TEXTURE);

As far as I understand, to make it YUV it's enough to set it to USAGE_HW_VIDEO_ENCODER, and perhaps that's why recorder mode doesn't work for some devices, what do you think? I don't have access to adaptation repo's so I can't build a hacked version, unfortunately.

Reply:
> As far as I understand, to make it YUV it's enough to set it to USAGE_HW_VIDEO_ENCODER, and perhaps that's why recorder mode doesn't work for some devices, what do you think? I don't have access to adaptation repo so I can't build a hacked version, unfortunately.

I can try it, but I'm not expecting different results. I think the usage bits are more hints to the driver about what memory allocation and locking strategies might work best for the usage and not all hints will apply to all buffer producers. So far as I understand the graphics buffers aren't really intended to be used for read back, that's what the data in the CAMERA_MSG_PREVIEW_FRAME and CAMERA_MSG_VIDEO_FRAME messages are for. It's unfortunate that droidmedia is abstracting away the reference counted lifetime of the data from the CAMERA_MSG_PREVIEW_FRAME  message.

The recording failure on the XA2 is probably a bug in the implementation for later android versions. As it's not used by gst-droid it has quite likely gone untested.

Reply:
So looking at logcat output the video recording problem in more recent android versions on some hardware at least is that only the  [`VIDEO_BUFFER_MODE_BUFFER_QUEUE`](https://android.googlesource.com/platform/frameworks/av/+/master/camera/include/camera/android/hardware/ICamera.h#48) mode is supported

`07-07 01:32:40.307  3076 23883 E Camera2Client: startRecordingL: Camera 0: Recording only supported buffer queue mode, but mode 0 is requested!`

Which can be enabled by calling `droid_media_camera_store_meta_data_in_buffers(.., true)` But then recording still fails because that requires that a [buffer queue be set as the video target](https://android.googlesource.com/platform/frameworks/av/+/master/camera/include/camera/android/hardware/ICamera.h#141) and nothing does that. It also means my understanding of the architecture's somewhat out of date.

Reply:
Thank you for trying this. Looks like the viewfinder frames on Sony devices are in some kind of proprietary format, so nothing left except to try to implement a video target for recorder mode.

Reply:
Changing the queue usage flag had no effect.  I also tried setting a queue as the video target, but that's still failing but now on an unsupported format error followed by some sort of lock up. I'm not sure what's happening there, the format is being left as the default which in the parameters is `video-frame-format=android-opaque` and changing that doesn't appear to have any effect.

Now that I've gone through the process myself; the repo manifests for the Sony device adaptions are in https://github.com/mer-hybris/android so everything you need to build it yourself should be publicly available.

Reply:
Thank you. I have exactly the same result on MediaTek device. For some unknown reason HAL chose format to be RGBA_8888, but then it reports that a scaler is missing for this format. Changing usage to `USAGE_HW_VIDEO_ENCODER` didn't help (however [here](http://androidxref.com/9.0.0_r3/xref/device/generic/goldfish/camera/EmulatedQemuCamera3.cpp#887) it has a key role), so I keep diving into Android code.

Reply:
It seems that the only way to get raw frames in a platform independent way is to use HAL_PIXEL_FORMAT_YCBCR_420_888 for the buffer queue and map buffers using [lockYCbCr](http://androidxref.com/9.0.0_r3/xref/frameworks/native/include/ui/GraphicBuffer.h#168). I updated my PR https://github.com/sailfishos/droidmedia/pull/81. I think it's too complex to add a proper support for this to gst-droid, so the next step is rewriting the capture code using plain droidmedia. As far as I understand, adding droidmedia headers to pj:oss is non-trivial, so the code will be split into a device-specific plugin and the plugin API will be hard-coded into gecko.

Reply:
There are definite advantages to having a plugin API, shorter build times if you're just making changes to it being my favorite.

But for the record, a hardware adaptation independent droidmedia-devel package should in theory be almost trivial. The issues I can identify are; You'd have to create a separate .spec file that manually installs the headers instead of using the Android makefile. There is an ifdef in the droidmedia constants header file that renames some members of a struct to follow name changes between Android versions which could be replaced with a union or just picking one of the names. And instead of linking against libhybris-common you would have to dlopen it and manually resolve the android_dlopen and android_dlsym symbols.

It's not a necessity, but there is a developing pattern of project specific plugins to utilize the glue code for loading for what is effectively a droidmedia plugin. If short cutting that is the same or less work than a new plugin API then I'd consider it.

Reply:
I was thinking specifically of gst-droid and nemo-gst-interfaces. Because the devel package of droidmedia is built per device, that means gst-droid is per device and can't provide its own devel package that applications can use. So we have this other package nemo-gst-interfaces that is device agnostic but exclusively contains gst-droid specific interfaces interfaces to act as a go between. And if you ever have reason to change one of those interfaces it's a three project minimum edit. And then we also have a plugin for qtmultimedia to act as an installable bridge between it and nemo-gst-interfaces.

That last point is something I overlooked earlier, not all devices will have droidmedia installed, and selectively installing a plugin for the browser is probably the best way to manage that. I'd still like to be able to build that plugin and gst-droid without a hardware specific sdk though.